File: words_counter.py
import re
import threading
import timeit
from concurrent.futures import ThreadPoolExecutor
from threading import Thread


def count_words(line):
    local_count = {}
    words = re.findall(r'\w+', line.rstrip("\n").lower())
    for counter in range(words.__len__()):
        word = words[counter]
        if word in local_count:
            local_count[word] += 1
        else:
            local_count[word] = 1

    with lock:
        for word, count in local_count.items():
            if word in word_count:
                word_count[word] += count
            else:
                word_count[word] = count

def count_word_occurrences(file_path):
    executor = ThreadPoolExecutor(max_workers=4)
    try:
        with open(file_path, 'r') as file:
            for line in file:
                executor.submit(count_words,line)

        return word_count
    except FileNotFoundError:
        print("File not found.")
    except IOError:
        print("An error occurred while reading the file.")

file_path = 'gnu_license.txt'
word_count = {}
lock = threading.Lock()

word_occurrences = count_word_occurrences(file_path)
for word, count in word_occurrences.items():
    print(word, " : ", count)

File: WordsCounter.java
package threads;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.regex.Pattern;

public class WordsCounter {

    public static void main(String[] args) {
        String filePath = System.getProperty("user.dir")+ "\\src\\threads\\gnu_license.txt";;
        long startTime = System.nanoTime();
        Map<String, Integer> wordOccurrences = countWordOccurrences(filePath, 4);
        wordOccurrences.forEach((word, count) -> System.out.println(word + ":" + count));
        long endTime = System.nanoTime();
        long duration = (endTime - startTime);
        System.out.println("Total time taken: "+ duration);
    }

    public static Map<String, Integer> countWordOccurrences(String filePath, int numThreads) {
        ConcurrentHashMap<String, Integer> wordCount = new ConcurrentHashMap<>();

        try (BufferedReader reader = new BufferedReader(new FileReader(filePath))) {

            String line;
            ExecutorService executorService = Executors.newFixedThreadPool(numThreads);
            while ((line = reader.readLine()) != null) {
                executorService.execute(new WordCountTask(line, wordCount));
            }

            executorService.shutdown();
            executorService.awaitTermination(1, TimeUnit.MINUTES);
        } catch (IOException e) {
            System.out.println("An error occurred while reading the file.");
        } catch (InterruptedException e) {
            System.out.println("Thread execution was interrupted.");
        }

        return wordCount;
    }

    static class WordCountTask implements Runnable {
        private final String contents;
        private final Map<String, Integer> wordCount;

        public WordCountTask(String contents,  Map<String, Integer> wordCount) {
            this.contents = contents;
            this.wordCount = wordCount;
        }

        @Override
        public void run() {
            String[] words = Pattern.compile("\\W+").splitAsStream(contents.toLowerCase())
                    .filter(word -> !word.isEmpty())
                    .toArray(String[]::new);
            for (int i = 0; i < words.length; i++) {
                wordCount.compute(words[i], (key, value) -> (value == null) ? 1 : value + 1);
            }
        }
    }
}
